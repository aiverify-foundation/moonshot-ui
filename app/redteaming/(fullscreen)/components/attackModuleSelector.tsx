import { IconName } from '@/app/components/IconSVG';
import { useGetAllAttackModulesQuery } from '@/app/services/attack-modules-api-service';
import { SelectListItem } from '@/app/views/shared-components/selectListItem';
import tailwindConfig from '@/tailwind.config';
const colors = tailwindConfig.theme?.extend?.colors as CustomColors;

type AttackModuleSelectViewProps = {
  selectedAttack?: AttackModule;
  onAttackClick: (attack: AttackModule) => void;
  onSkipClick?: () => void;
};

function AttackModuleSelectView(props: AttackModuleSelectViewProps) {
  const { selectedAttack, onAttackClick } = props;
  const { data, isFetching } = useGetAllAttackModulesQuery();

  return (
    <div className="flex flex-col pt-4 gap-8 pb-4 h-[80%] items-center">
      <section className="flex flex-col items-center gap-3">
        <hgroup>
          <h2 className="text-[1.6rem] leading-[2rem] height-[2rem] font-medium tracking-wide text-white w-full text-center">
            Would you like to use any of these attack modules?
          </h2>
          <p className="text-[0.9rem] text-moongray-400 text-center">
            <span className="font-bold">Optional</span>&nbsp;modules that
            automatically generate prompts to red team an Endpoint.
          </p>
        </hgroup>
      </section>
      <div className="relative flex flex-col min-h-[300px] px-[10%] w-[100%] h-full items-center">
        {isFetching || !data ? (
          <div className="ring">
            Loading
            <span />
          </div>
        ) : (
          <>
            <ul
              className="flex flex-row flex-wrap gap-[2%] w-[100%] overflow-y-auto custom-scrollbar px-4"
              style={{ height: '100%' }}>
              {data.map((attack) => {
                const isSelected =
                  selectedAttack && selectedAttack.id === attack.id;
                const description = attack.description
                  ? attack.description
                      .split('Parameters:')[0]
                      .replace(/\n/g, '<br />')
                  : '';
                return (
                  <SelectListItem<AttackModule>
                    key={attack.id}
                    label={attack.name}
                    onClick={() => onAttackClick(attack)}
                    iconName={IconName.OutlineBox}
                    item={attack}
                    hideCheckbox
                    bgColor={isSelected ? colors.moonwine[950] : undefined}
                    style={{
                      alignItems: 'flex-start',
                      height: 180,
                      overflowY: 'hidden',
                    }}>
                    <p
                      className="text-[0.8rem] text-moongray-400"
                      dangerouslySetInnerHTML={{ __html: description }}
                    />
                  </SelectListItem>
                );
              })}
            </ul>
          </>
        )}
      </div>
    </div>
  );
}

export { AttackModuleSelectView };
