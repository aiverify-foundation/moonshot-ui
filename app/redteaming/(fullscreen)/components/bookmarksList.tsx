function BookmarksList() {
  return (
    <div className="flex flex-col min-h-[300px]">
      <h3 className="mb-5 text-white">Bookmarks</h3>
      <ul className="list-none bg-white rounded-sm">
        <li className="p-1 border-b">
          <p>Bookmark 1</p>
        </li>
        <li className="p-1 border-b">
          <p>Bookmark 2</p>
        </li>
        <li className="p-1 border-b">
          <p>Bookmark 3</p>
        </li>
        <li className="p-1">
          <p>Bookmark 4</p>
        </li>
      </ul>
    </div>
  );
}

export { BookmarksList };
