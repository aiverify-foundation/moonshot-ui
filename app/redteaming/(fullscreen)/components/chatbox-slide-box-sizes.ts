export type SlideChatBoxDimensions = {
  width: number;
  height: number;
  gap: number;
};

export function getDefaultChatBoxSizes() {
  // IMPORTANT: Also align the values of --chatwindow-width and --gap-width css variables in global.css

  // Default values
  let width = 420;
  let height = 550;
  let gap = 16;

  if (typeof window === 'undefined') {
    return { width: 0, height: 0, gap: 0 };
  }

  // Media query for large screens (e.g., more than 1180px)
  if (window.matchMedia('(min-width: 1180px)').matches) {
    console.log('1180px');
    width = 420; // Smaller width for small screens
    height = 500; // Smaller height for small screens
    gap = 50; // Smaller gap for small screens
  }

  // Media query for medium screens (e.g., less than 992px)
  if (window.matchMedia('(max-width: 992px)').matches) {
    console.log('992px');
    width = 420; // Smaller width for small screens
    height = 450; // Smaller height for small screens
    gap = 15; // Smaller gap for small screens
  }

  // Media query for small screens (e.g., less than 768px)
  if (window.matchMedia('(max-width: 768px)').matches) {
    console.log('768px');
    width = 300; // Smaller width for small screens
    height = 350; // Smaller height for small screens
    gap = 15; // Smaller gap for small screens
  }

  // Media query for very small screens (e.g., less than 480px)
  if (window.matchMedia('(max-width: 480px)').matches) {
    console.log('480px');
    width = 280; // Even smaller width for very small screens
    height = 300; // Even smaller height for very small screens
    gap = 10; // Even smaller gap for very small screens
  }

  // Return updated values
  return { width, height, gap };
}
