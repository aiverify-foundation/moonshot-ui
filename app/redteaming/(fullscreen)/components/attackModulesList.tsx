import React from 'react';
import { Icon, IconName } from '@/app/components/IconSVG';
import { Button, ButtonType } from '@/app/components/button';
import { useGetAllAttackModulesQuery } from '@/app/services/attack-modules-api-service';
import { colors } from '@/app/views/shared-components/customColors';
import { LoadingAnimation } from '@/app/views/shared-components/loadingAnimation';

type AttackModulesListProps = {
  onPrimaryBtnClick: (attackModule: AttackModule) => void;
  onSecondaryBtnClick?: () => void;
};

function AttackModulesList(props: AttackModulesListProps) {
  const { onPrimaryBtnClick, onSecondaryBtnClick } = props;
  const [selectedAttackModule, setSelectedAttackModule] = React.useState<
    AttackModule | undefined
  >();
  const { data, refetch, isLoading } = useGetAllAttackModulesQuery();

  React.useEffect(() => {
    refetch();
  }, []);

  let attackModuleDetailsSection: React.ReactNode;

  if (selectedAttackModule) {
    const description = selectedAttackModule.description
      ? selectedAttackModule.description.replace(/\n/g, '<br /><br/>')
      : '';

    attackModuleDetailsSection = (
      <section className="p-2 bg-moongray-950 w-[480px]">
        <header className="flex flex-col gap-2 mb-4">
          <div className="flex gap-2">
            <Icon name={IconName.MoonAttackStrategy} />
            <h3 className="text-[1rem] font-semibold text-white">
              {selectedAttackModule.name}
            </h3>
          </div>
        </header>
        <p
          dangerouslySetInnerHTML={{ __html: description }}
          className="text-[0.85rem] overflow-y-auto custom-scrollbar h-full text-moongray-200"
          style={{ height: 'calc(100% - 40px)' }}
        />
      </section>
    );
  }

  const listOfModules = data ? (
    <ul className="divide-y divide-moongray-500 max-w-[400px] overflow-x-hidden overflow-y-auto custom-scrollbar h-full">
      {data.map((attackModule) => {
        const isSelected =
          selectedAttackModule && selectedAttackModule.id === attackModule.id;
        const description = attackModule.description
          ? attackModule.description
              .split('Parameters:')[0]
              .replace(/\n/g, '<br />')
          : '';
        return (
          <li
            key={attackModule.id}
            className={`p-2 ${!isSelected && 'hover:bg-moongray-900'} active:bg-moongray-600 cursor-pointer 
            ${isSelected ? 'bg-moongray-950' : 'bg-moongray-700'}`}
            onClick={() => setSelectedAttackModule(attackModule)}
            style={{
              transition: 'background-color 0.2s ease-in-out',
            }}>
            <header className="flex flex-col gap-2">
              <div className="flex gap-2">
                <Icon name={IconName.MoonAttackStrategy} />
                <h3 className="text-[0.87rem] text-white">
                  {attackModule.name}
                </h3>
              </div>
              <p
                className="text-sm h-[80px] min-h-[80px] overflow-y-hidden text-[0.74rem] text-moongray-200"
                dangerouslySetInnerHTML={{ __html: description }}
              />
            </header>
          </li>
        );
      })}
    </ul>
  ) : null;

  return (
    <div className="relative flex flex-col w-full h-full gap-4">
      {isLoading || !data ? (
        <LoadingAnimation />
      ) : (
        <>
          <main
            className="h-full"
            style={{ height: 'calc(100% - 70px)' }}>
            <div className="flex gap-4 h-full">
              {listOfModules}
              {attackModuleDetailsSection}
            </div>
          </main>
          <footer className="flex justify-end gap-2">
            <Button
              mode={ButtonType.OUTLINE}
              onClick={onSecondaryBtnClick}
              text="Cancel"
              hoverBtnColor={colors.moongray[700]}
              pressedBtnColor={colors.moongray[800]}
            />
            <Button
              disabled={selectedAttackModule === undefined}
              mode={ButtonType.PRIMARY}
              onClick={() =>
                selectedAttackModule
                  ? onPrimaryBtnClick(selectedAttackModule)
                  : null
              }
              text="Use"
              hoverBtnColor={colors.moongray[950]}
            />
          </footer>
        </>
      )}
    </div>
  );
}

export { AttackModulesList };
