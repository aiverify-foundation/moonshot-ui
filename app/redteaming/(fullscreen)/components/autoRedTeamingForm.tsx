import { ChangeEvent, useState } from 'react';
import { Button, ButtonType } from '@/app/components/button';
import { SelectInput, SelectOption } from '@/app/components/selectInput';
import { TextArea } from '@/app/components/textArea';
import { useGetAllRecipesQuery } from '@/app/services/recipe-api-service';
import tailwindConfig from '@/tailwind.config';

const colors = tailwindConfig.theme?.extend?.colors as CustomColors;

function AutoRedTeamingForm() {
  const [selectedRecipe, setSelectedRecipe] = useState<Recipe | undefined>(
    undefined
  );
  const [initialPrompt, setInitialPrompt] = useState<string>('');

  const { data: recipes, isLoading: recipesIsLoading } =
    useGetAllRecipesQuery(undefined);

  let recipesSelectionOptions: SelectOption[] = [];
  if (recipes && recipes.length) {
    recipesSelectionOptions = recipes.map((rcp) => ({
      value: rcp.id,
      label: rcp.name,
    }));
  }

  function handleChange(
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    if (e.target.name === 'recipe') {
      setSelectedRecipe(recipes?.find((rcp) => rcp.id === e.target.value));
    } else if (e.target.name === 'initialPrompt') {
      setInitialPrompt(e.target.value);
    }
  }

  return (
    <div className="relative flex flex-col min-h-[300px]">
      <h3 className="mb-5 text-white">Automated Red Teaming</h3>
      {recipesIsLoading ? (
        <div className="ring">
          Loading
          <span />
        </div>
      ) : (
        <>
          <SelectInput
            label="Select Recipe"
            name="recipe"
            value={selectedRecipe?.id}
            placeholder="Select a recipe"
            options={recipesSelectionOptions}
            onSyntheticChange={handleChange}
          />
          <TextArea
            name="initialPrompt"
            label="Initial Prompt"
            onChange={handleChange}
            value={initialPrompt}
            placeholder="Enter initial prompt here"
          />
          <div className="flex justify-left mt-2">
            <Button
              type="submit"
              mode={ButtonType.PRIMARY}
              text="Start Red Team"
              btnColor={colors.moongray[950]}
              hoverBtnColor={colors.moongray[900]}
              textColor={colors.white}
            />
          </div>
        </>
      )}
    </div>
  );
}

export { AutoRedTeamingForm };
