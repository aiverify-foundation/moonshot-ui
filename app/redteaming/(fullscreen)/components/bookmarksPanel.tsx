import React from 'react';
import { IconName } from '@/app/components/IconSVG';
import { Button, ButtonType } from '@/app/components/button';
import { colors } from '@/app/customColors';
import { ViewBookmarksModal } from './viewBookmarksModal';

function BookmarksPanel({
  bottom,
  left,
  disabled,
  defaultShowPanel = false,
  onUseBtnClick,
  onPanelClose,
}: {
  bottom: React.CSSProperties['top'];
  left: React.CSSProperties['left'];
  disabled: boolean;
  defaultShowPanel?: boolean;
  onPanelClose?: () => void;
  onUseBtnClick: (preparedPrompt: string) => void;
}) {
  const [showPanel, setShowPanel] = React.useState(() => defaultShowPanel);

  React.useEffect(() => {
    setShowPanel(defaultShowPanel);
  }, [defaultShowPanel]);

  function handleUseBtnClick(preparedPrompt: string) {
    onUseBtnClick(preparedPrompt);
    setShowPanel(false);
    if (onPanelClose) {
      onPanelClose();
    }
  }

  function handleCancelBtnClick() {
    setShowPanel(false);
    if (onPanelClose) {
      onPanelClose();
    }
  }

  return (
    <div
      className="bg-moongray-600 w-[220px] absolute rounded-md p-2 shadow-lg
        flex gap-4"
      style={{ bottom, left }}>
      {disabled && (
        <div
          className="absolute gap-2 bg-moongray-950/50 w-full h-full z-10
            flex justify-center items-center rounded-md"
          style={{ top: 0, left: 0 }}>
          <div className="waitspinner" />
        </div>
      )}
      {showPanel && (
        <ViewBookmarksModal
          onCloseIconClick={handleCancelBtnClick}
          onPrimaryBtnClick={handleUseBtnClick}
        />
      )}
      <div className="flex items-center gap-2">
        <Button
          width={200}
          alignContent="flex-start"
          text="Bookmarked Prompts"
          type="button"
          mode={ButtonType.TEXT}
          hoverBtnColor={colors.moongray[500]}
          pressedBtnColor={colors.moongray[400]}
          leftIconName={IconName.Ribbon}
          onClick={() => setShowPanel(true)}
        />
      </div>
    </div>
  );
}

export { BookmarksPanel };
