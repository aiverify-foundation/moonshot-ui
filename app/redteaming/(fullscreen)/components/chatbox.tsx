import React, { useState } from 'react';
import { Tooltip, TooltipPosition } from '@/app/components/tooltip';
import { colors } from '@/app/views/shared-components/customColors';
import { PromptBubbleInfo } from './prompt-bubble-info';
import { Chat } from '@components/chat';

type ChatBoxProps = {
  windowId: string;
  title: string;
  resizable: boolean;
  draggable: boolean;
  disableCloseIcon?: boolean;
  disableOnScroll: boolean;
  chatHistory: DialoguePairInfo[];
  disableBubbleTooltips?: boolean;
  initialXY: [number, number];
  initialSize: [number, number];
  initialScrollTop: number;
  promptTemplates: PromptTemplate[];
  currentPromptTemplate?: PromptTemplate;
  currentPromptText?: string;
  isAttackMode: boolean;
  isChatCompletionInProgress: boolean;
  styles?: React.CSSProperties;
  headerStyle?: React.CSSProperties;
  resizeHandlerColor?: React.CSSProperties['color'];
  onWindowChange: (
    x: number,
    y: number,
    width: number,
    height: number,
    scrollPosition: number,
    windowId: string
  ) => void;
  onWheel?: (event: React.WheelEvent<HTMLDivElement>) => void;
  onCloseClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onWholeWindowClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
};

export type ChatBoxControls = {
  scrollToBottom: () => void;
  scrollToTop: () => void;
};

const ChatBox = React.forwardRef(
  (props: ChatBoxProps, ref: React.Ref<ChatBoxControls>) => {
    const {
      windowId,
      resizable,
      draggable,
      disableCloseIcon = true,
      disableOnScroll,
      disableBubbleTooltips = false,
      title,
      chatHistory,
      initialXY,
      initialSize,
      initialScrollTop,
      promptTemplates,
      currentPromptTemplate,
      currentPromptText,
      isAttackMode = false,
      isChatCompletionInProgress,
      styles,
      headerStyle,
      resizeHandlerColor,
      onWindowChange,
      onWheel,
      onCloseClick,
      onWholeWindowClick,
    } = props;

    const [dialoguePairHovered, setDialoguePairHovered] = useState<
      number | undefined
    >();

    const scrollDivRef = React.useRef<HTMLDivElement>(null);
    React.useImperativeHandle(ref, () => ({
      scrollToBottom,
      scrollToTop,
    }));

    function scrollToBottom() {
      if (scrollDivRef.current) {
        scrollDivRef.current.scrollTop = scrollDivRef.current.scrollHeight;
      }
    }

    function scrollToTop() {
      if (scrollDivRef.current) {
        scrollDivRef.current.scrollTop = scrollDivRef.current.scrollHeight;
      }
    }

    React.useLayoutEffect(() => {
      scrollToBottom();
    }, [chatHistory]);

    return (
      <Chat.Container
        ref={scrollDivRef}
        windowId={windowId}
        name={title}
        initialXY={initialXY}
        initialSize={initialSize}
        initialScrollTop={initialScrollTop}
        resizable={resizable}
        draggable={draggable}
        disableOnScroll={disableOnScroll}
        disableCloseIcon={disableCloseIcon}
        styles={styles}
        headerStyle={headerStyle}
        resizeHandlerColor={resizeHandlerColor}
        onWindowChange={onWindowChange}
        onWheel={onWheel}
        onCloseClick={onCloseClick}
        onWholeWindowClick={onWholeWindowClick}>
        {chatHistory.map((dialogue, index) => {
          const appliedPromptTemplate = promptTemplates
            ? promptTemplates.find(
                (template) => template.name === dialogue.prompt_template
              )
            : undefined;
          return (
            <li
              className="flex flex-col p-2"
              key={index}
              onMouseEnter={() => setDialoguePairHovered(index)}
              onMouseLeave={() => setDialoguePairHovered(undefined)}>
              <div className="flex flex-col text-right pr-2 text-sm text-white">
                You
              </div>
              <div className="self-end snap-top max-w-[90%]">
                <div className="flex items-center">
                  <Tooltip
                    disabled={disableBubbleTooltips}
                    position={TooltipPosition.top}
                    contentMaxWidth={500}
                    contentMinWidth={300}
                    backgroundColor={colors.moongray[700]}
                    fontColor={colors.white}
                    offsetTop={-10}
                    content={
                      <PromptBubbleInfo
                        duration={dialogue.duration}
                        prompt={dialogue.prepared_prompt}
                        prompt_template={dialogue.prompt_template}
                        templateString={
                          appliedPromptTemplate
                            ? appliedPromptTemplate.template
                            : undefined
                        }
                      />
                    }>
                    <Chat.TalkBubble
                      backgroundColor={colors.imdapurple}
                      fontColor="#FFF"
                      styles={{
                        marginBottom: 0,
                        fontSize: 14,
                        border:
                          dialoguePairHovered === index
                            ? '1px solid #2563eb'
                            : '1px solid transparent',
                      }}>
                      {dialogue.prepared_prompt}
                    </Chat.TalkBubble>
                  </Tooltip>
                </div>
              </div>
              <div className="max-w-[90%] flex flex-col text-left pl-2 text-sm text-white">
                AI
              </div>
              <Chat.TalkBubble
                backgroundColor={colors.chatBubbleWhite}
                fontColor={colors.black}
                styles={{
                  fontSize: 14,
                  textAlign: 'left',
                  border:
                    dialoguePairHovered === index
                      ? '1px solid #2563eb'
                      : '1px solid transparent',
                }}>
                {dialogue.predicted_result}
              </Chat.TalkBubble>
            </li>
          );
        })}

        {isChatCompletionInProgress && !isAttackMode && (
          <div className="flex flex-col p-2">
            <div className="flex flex-col text-right pr-2 text-xs text-white">
              You
            </div>
            <Chat.TalkBubble
              backgroundColor={colors.imdapurple}
              fontColor="#FFF"
              styles={{ alignSelf: 'flex-end', fontSize: 14 }}>
              {currentPromptTemplate && currentPromptText
                ? currentPromptTemplate.template.replace(
                    '{{ prompt }}',
                    currentPromptText
                  )
                : currentPromptText}
            </Chat.TalkBubble>
            <div className="flex flex-col text-left pl-2 text-xs text-white">
              AI
            </div>
            <div className="flex justify-start mr-4">
              <Chat.LoadingAnimation
                backgroundColor={colors.white}
                dotColor={colors.imdapurple}
              />
            </div>
          </div>
        )}

        {isChatCompletionInProgress && isAttackMode && (
          <div className="flex justify-start mr-4 p-2">
            <Chat.LoadingAnimation
              backgroundColor={colors.white}
              dotColor={colors.imdapurple}
            />
          </div>
        )}
      </Chat.Container>
    );
  }
);

ChatBox.displayName = 'Chatbox';

export { ChatBox };
