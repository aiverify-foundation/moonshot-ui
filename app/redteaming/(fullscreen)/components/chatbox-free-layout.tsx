import React, { useEffect, useState } from 'react';
import { colors } from '@/app/customColors';
import {
  getWindowId,
  getWindowScrollTopById,
  getWindowSizeById,
  getWindowXYById,
} from '@/app/lib/window-utils';
import useChatboxesPositionsUtils from '@/app/redteaming/(fullscreen)/hooks/useChatboxesPositionsUtils';
import { useAppSelector } from '@/lib/redux';
import { ChatBox, ChatBoxControls } from './chatbox';

const minimizedStyle = {
  transform: 'scale(0.1)',
  width: 500,
  height: 500,
  top: `calc(100vh - 350px)`,
  transition:
    'transform 0.3s ease-in, top 0.4s ease-in-out, left 0.5s ease-out', // Smooth transition for the animation
};

type ChatFreeLayoutProps = {
  chatSession: SessionData;
  chatCompletionInProgress: boolean;
  isAttackMode: boolean;
  promptTemplates: PromptTemplate[];
  selectedPromptTemplate: PromptTemplate | undefined;
  promptText: string;
  handleOnWindowChange: (
    x: number,
    y: number,
    width: number,
    height: number,
    scrollPosition: number,
    windowId: string
  ) => void;
  handleCreatePromptBookmarkClick: CreatePromptBookmarkFunction;
};

const ChatboxFreeLayout = React.forwardRef(
  (
    props: ChatFreeLayoutProps,
    ref: React.Ref<Map<string, ChatBoxControls>>
  ) => {
    const {
      chatSession,
      chatCompletionInProgress,
      isAttackMode,
      promptTemplates,
      selectedPromptTemplate,
      promptText,
      handleOnWindowChange,
      handleCreatePromptBookmarkClick,
    } = props;
    const windowsMap = useAppSelector((state) => state.windows.map);
    const { resetChatboxPositions } = useChatboxesPositionsUtils(chatSession);
    const [minizedChats, setMinizedChats] = useState<string[]>([]);
    const [isMaximizing, setIsMazimizing] = useState(false);
    const chatBoxControlsMap = new Map<string, ChatBoxControls>();
    React.useImperativeHandle(ref, () => chatBoxControlsMap);

    function handleMinimizeClick(windowId: string) {
      return () => {
        setMinizedChats([...minizedChats, windowId]);
      };
    }

    function handleMaximizeClick(windowId: string) {
      return () => {
        setIsMazimizing(true);
        setMinizedChats(minizedChats.filter((id) => id !== windowId));
      };
    }

    useEffect(() => {
      resetChatboxPositions(false);
    }, []);

    useEffect(() => {
      setTimeout(() => {
        setIsMazimizing(false);
      }, 300);
    }, [isMaximizing]);

    const chatboxHeaderStyle: React.CSSProperties = {
      borderTopLeftRadius: '0.5rem',
      borderTopRightRadius: '0.5rem',
    };

    return (
      <>
        {chatSession.session.endpoints.map((id: string, index: number) => {
          const isMinimized = minizedChats.includes(getWindowId(id));
          const left = index * 20;
          const chatboxStyle = isMinimized
            ? { ...minimizedStyle, left }
            : isMaximizing
              ? {
                  transition:
                    'transform 0.3s ease-in-out, top 0.5s ease-in-out, left 0.5s ease-in-out',
                }
              : { borderRadius: '0.5rem' };
          return windowsMap[getWindowId(id)] ? (
            <ChatBox
              key={id}
              disableCloseIcon={
                chatSession.session.endpoints.length === 1 ||
                chatSession.session.endpoints.length - minizedChats.length === 1
              }
              resizable
              draggable
              disableOnScroll
              disableBubbleTooltips={isMinimized}
              resizeHandlerColor={colors.moonwine[500]}
              ref={(imperativeHandles) => {
                if (imperativeHandles) {
                  chatBoxControlsMap.set(id, imperativeHandles);
                } else {
                  chatBoxControlsMap.delete(id);
                }
              }}
              windowId={getWindowId(id)}
              title={id}
              initialXY={getWindowXYById(windowsMap, id)}
              initialSize={getWindowSizeById(windowsMap, id)}
              initialScrollTop={getWindowScrollTopById(windowsMap, id)}
              chatHistory={
                chatSession.chat_records
                  ? chatSession.chat_records[id] || []
                  : []
              }
              promptTemplates={promptTemplates}
              currentPromptTemplate={selectedPromptTemplate}
              isAttackMode={isAttackMode}
              currentPromptText={promptText}
              isChatCompletionInProgress={chatCompletionInProgress}
              onWindowChange={handleOnWindowChange}
              styles={chatboxStyle}
              headerStyle={chatboxHeaderStyle}
              onCloseClick={handleMinimizeClick(getWindowId(id))}
              onWholeWindowClick={
                isMinimized ? handleMaximizeClick(getWindowId(id)) : undefined
              }
              onCreatePromptBookmarkClick={handleCreatePromptBookmarkClick}
            />
          ) : null;
        })}
      </>
    );
  }
);

ChatboxFreeLayout.displayName = 'ChatboxFreeLayout';

export { ChatboxFreeLayout };
