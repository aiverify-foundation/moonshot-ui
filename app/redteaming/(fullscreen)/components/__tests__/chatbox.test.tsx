import { render, screen } from '@testing-library/react';
import { ChatBox } from '@/app/redteaming/(fullscreen)/components/chatbox';
import { useAppDispatch, useAppSelector } from '@/lib/redux';

jest.mock('@/lib/redux', () => ({
  useAppDispatch: jest.fn(),
  useAppSelector: jest.fn(),
}));

const mockPromptTemplates: PromptTemplate[] = [
  {
    name: 'template1',
    description: 'Description for Template 1',
    template: 'This is the template for {{ prompt }} with some context',
  },
  {
    name: 'template2',
    description: 'Description for Template 2',
    template: 'Another template with {{ prompt }} and additional information',
  },
];

const mockPromptDetails: PromptDetails[] = [
  {
    chat_record_id: 1,
    conn_id: 'conn1',
    context_strategy: 'strategy1',
    prompt_template: 'template1',
    attack_module: 'module1',
    metric: 'metric1',
    prompt: 'Test prompt 1',
    prepared_prompt: 'Prepared test prompt 1',
    system_prompt: 'System prompt 1',
    predicted_result: 'Predicted result 1',
    duration: '1s',
    prompt_time: '2023-06-01T12:00:00Z',
  },
  {
    chat_record_id: 2,
    conn_id: 'conn2',
    context_strategy: 'strategy2',
    prompt_template: 'template2',
    attack_module: 'module2',
    metric: 'metric2',
    prompt: 'Test prompt 2',
    prepared_prompt: 'Prepared test prompt 2',
    system_prompt: 'System prompt 2',
    predicted_result: 'Predicted result 2',
    duration: '2s',
    prompt_time: '2023-06-01T12:01:00Z',
  },
];

describe('Chatbox', () => {
  beforeAll(() => {
    const mockDispatch = jest.fn();
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useAppSelector as jest.Mock).mockReturnValue(jest.fn());
  });
  it('render chatbox containing prompt and response talk bubbles', () => {
    const { container } = render(
      <ChatBox
        windowId="test-window"
        title="Test Chat"
        resizable={false}
        draggable={false}
        disableOnScroll={true}
        chatHistory={mockPromptDetails}
        initialXY={[0, 0]}
        initialSize={[300, 400]}
        initialScrollTop={0}
        promptTemplates={mockPromptTemplates}
        isAttackMode={false}
        isChatCompletionInProgress={false}
        onWindowChange={() => null}
        onCreatePromptBookmarkClick={() => null}
      />
    );

    expect(
      screen.getByText(mockPromptDetails[0].prepared_prompt)
    ).toBeInTheDocument();
    expect(
      screen.getByText(mockPromptDetails[0].predicted_result)
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
