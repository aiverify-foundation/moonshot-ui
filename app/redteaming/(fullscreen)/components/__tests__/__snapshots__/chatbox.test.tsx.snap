// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chatbox render chatbox containing prompt and response talk bubbles 1`] = `
<div>
  <div
    class="absolute pt-0 text-white 
          min-w-96 shadow-moongray-800
          dark:shadow-moongray-900/30 bg-moongray-950 backdrop-blur-sm 
          
        "
    id="test-window"
    style="left: 0px; top: 0px; background-color: rgb(45, 43, 47); width: 300px; height: 400px;"
  >
    <div
      class="flex flex-col w-full h-full"
    >
      <div
        class="flex flex-col w-full mb-6 bg-moongray-1000"
        style="margin-bottom: 14px;"
      >
        <div
          class="flex px-3 justify-between w-full"
        >
          <div
            class="flex items-center h-7 text-sm"
          >
            Test Chat
          </div>
        </div>
      </div>
      <div
        class="h-full px-4 overflow-x-hidden overflow-y-auto   bg-white size-full custom-scrollbar snap-mandatory"
        style="padding-left: 0px; padding-right: 0px; margin-left: 0.7rem; margin-right: 0.7rem;"
      >
        <div
          class="h-full overflow-y-auto custom-scrollbar bg-chatboxbg"
          id="chatContainer"
        >
          <li
            class="flex flex-col p-2"
          >
            <h1
              class="flex flex-col text-right pr-2 text-sm text-white"
            >
              You
            </h1>
            <div
              class="self-end snap-top max-w-[90%]"
            >
              <div
                class="flex items-center"
              >
                <div
                  class="flex gap-2"
                >
                  <div
                    class="cursor-pointer pt-2"
                  >
                    <div
                      class="childWrapper"
                    >
                      <div
                        class="
        flex items-center justify-center 
        cursor-pointer
        hover:opacity-50 active:opacity-25
        
      "
                      >
                        <svg
                          fill="#FFFFFF"
                          height="20"
                          viewBox="0 0 14 19"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M7.75433 13.9464L7.40078 13.5929L7.04723 13.9464L3.00377 17.9899C2.04707 17.8913 1.30078 17.0827 1.30078 16.1V2.9C1.30078 1.85074 2.15152 1 3.20078 1H11.6008C12.65 1 13.5008 1.85074 13.5008 2.9V16.1C13.5008 17.0827 12.7545 17.8913 11.7978 17.9899L7.75433 13.9464Z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    class="childWrapper"
                  >
                    <div
                      style="text-align: right; color: rgb(255, 255, 255); padding: 12px 16px; font-size: 14px; background-color: rgb(73, 48, 90); margin: 0px 0px 0px 0px; border-radius: 14px; min-width: 35%; border: 1px solid transparent;"
                    >
                      Prepared test prompt 1
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <h1
              class="max-w-[90%] flex flex-col text-left pl-2 text-sm text-white"
            >
              AI
            </h1>
            <div
              style="text-align: left; color: rgb(0, 0, 0); padding: 12px 16px; font-size: 14px; background-color: rgb(221, 221, 221); margin: 0px 0px 0px 0px; border-radius: 14px; min-width: 35%; border: 1px solid transparent;"
            >
              Predicted result 1
            </div>
          </li>
          <li
            class="flex flex-col p-2"
          >
            <h1
              class="flex flex-col text-right pr-2 text-sm text-white"
            >
              You
            </h1>
            <div
              class="self-end snap-top max-w-[90%]"
            >
              <div
                class="flex items-center"
              >
                <div
                  class="flex gap-2"
                >
                  <div
                    class="cursor-pointer pt-2"
                  >
                    <div
                      class="childWrapper"
                    >
                      <div
                        class="
        flex items-center justify-center 
        cursor-pointer
        hover:opacity-50 active:opacity-25
        
      "
                      >
                        <svg
                          fill="#FFFFFF"
                          height="20"
                          viewBox="0 0 14 19"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M7.75433 13.9464L7.40078 13.5929L7.04723 13.9464L3.00377 17.9899C2.04707 17.8913 1.30078 17.0827 1.30078 16.1V2.9C1.30078 1.85074 2.15152 1 3.20078 1H11.6008C12.65 1 13.5008 1.85074 13.5008 2.9V16.1C13.5008 17.0827 12.7545 17.8913 11.7978 17.9899L7.75433 13.9464Z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <div
                    class="childWrapper"
                  >
                    <div
                      style="text-align: right; color: rgb(255, 255, 255); padding: 12px 16px; font-size: 14px; background-color: rgb(73, 48, 90); margin: 0px 0px 0px 0px; border-radius: 14px; min-width: 35%; border: 1px solid transparent;"
                    >
                      Prepared test prompt 2
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <h1
              class="max-w-[90%] flex flex-col text-left pl-2 text-sm text-white"
            >
              AI
            </h1>
            <div
              style="text-align: left; color: rgb(0, 0, 0); padding: 12px 16px; font-size: 14px; background-color: rgb(221, 221, 221); margin: 0px 0px 0px 0px; border-radius: 14px; min-width: 35%; border: 1px solid transparent;"
            >
              Predicted result 2
            </div>
          </li>
        </div>
      </div>
      <div
        class="flex items-center justify-start text-xs text-white/70 px-3"
        style="height: 17px;"
      />
    </div>
  </div>
</div>
`;
